### DISCUSSION TITLE
How can I best monitor the execution of each migration in 3.x when we previously used the Announcer functionality in 1.4.0

### DISCUSSION THREAD

**Discussion by jcolebrand** [fluentmigrator/fluentmigrator#1749] [Q&A] [answered]
**Describe the question**
With 1.4.0 we used a custom IAnnouncer (code after body for further explanation) and we could capture all the data coming out of the migrations as they happened.

As best I can follow, this is all deprecated behavior so I should shift my paradigm. (If this is where I'm wrong, I can probably get away with registering my own IAnnouncer implementation and that's that)

What we were tracking before is each migration, that it applied successfully, what was applied, etc, and then we logged that out as a final big bulk thing on our TeamCity logs, but on local machines we just dumped the relevant bits.

Now with IMigrationRunner.MigrateUp, I'm not sure I follow how to capture that information except by using a custom logger that looks for specific triggers or the like.

**Documentation Pages You've Read So Far**
Honestly I've cloned the repo down for both 1.4.0 and 3.3.1 and I see where you've marked a ton of things deprecated and I think I have a handle on it but this is a big project to keep in the head, yeah? Especially comparing two vastly different codebases at this point. I've read a ton of posts and other people's questions, and never kept track, but clearly I may have missed an obvious answer somewhere.

**Expected benefit**
I would like to capture details in memory of each migration as a tiny encapsulation of what happened _as well as_ the migration sequence number and SQL that was attempted to be applied, and if it errored or was successful, if it was up or down, etc. (This is so we can sort the errors and log concise recaps at the end, etc. Yay business needs.)

I hope the code to follow is helpful in understanding what we were doing before.

**Information (please complete the following information):**
 - OS: Windows 10/2016/2019
 - Platform: .Net Core 3.1.x
 - FluentMigrator version: 3.3.1
 - FluentMigrator runner: In process, similar to [https://github.com/fluentmigrator/fluentmigrator/blob/master/samples/FluentMigrator.Example.Migrator/Program.DependencyInjection.cs] but with a bit more plumbing around the edges :D
 - Database Management System: SQLServer, potentially others
 - Database Management System Version: Sql Server Enterprise, Developer

**Additional context**

First: I'm using a custom ILoggerFactory and ILogger due to the rest of my setup and all the things it has to do (fetch packages, extract things, do validation, etc etc) with one of two loggers injected in Startup.cs, so I wasn't really planning on using `.AddLogging(lb => lb.AddFluentMigratorConsole())` because as per [https://github.com/fluentmigrator/fluentmigrator/issues/930] so long as I'm already supplying the ILogger etc, you don't need me to register yours (besides, I want to intercept the logs, not write them to console only, etc etc, yay business needs). I can absolutely play around with the ILogger implementation I have, but I would rather keep those implementations cleaner, since they are just receiving messages and then doing the appropriate thing based on their implementation, as a good logger should.

Here are some selected files of how we were doing this in 1.4.0 FluentMigrator with an in-process runner

1.4.0 MigrationAnnouncer
```
using System;
using System.Linq;
using FluentMigrator.Runner;

namespace Our.Internal.Models
{
    /// <summary>
    /// The purpose of a MigrationAnnouncer is to print things appropriately during runtime.
    /// </summary>
    public class MigrationAnnouncer : IAnnouncer
    {
        private readonly MigrationResult _result;
        private Migration _migration = null;

        public MigrationAnnouncer(MigrationResult result)
        {
            this._result = result;
        }

        public void Heading(string message)
        {
            // We use an accumulator pattern so we can always modify the same migration in flight across many methods
            if (_migration != null)
            {
                _result.Migrations.Add(_migration);
            }

            // This indicates a new migration
            string[] split = message.Split(new[] { ' ', ':' }, StringSplitOptions.RemoveEmptyEntries); // Should have 3

            if (split.Length != 3)
            {
                // this indicates that we are not looking at <internal> migration, because <internal> migrations always have a triple-part
                // This is the migration sequence number (date-time string), the migration description, and the status of the migration (up/down)

                _migration = new Migration
                {
                    Description = message,
                    MigrationType = MigrationType.Unknown
                };
            }
            else
            {
                _migration = new Migration
                {
                    Description = split.Skip(1).First(),
                    MigrationType = string.Equals(split.Last(), "reverting", StringComparison.InvariantCultureIgnoreCase) ? MigrationType.Down : MigrationType.Up
                };
            }
        }

        public void Say(string message)
        {
            _migration.Say.Add(message);
        }

        public void Emphasize(string message)
        {
            _migration.Emphasize.Add(message);
        }

        public void Sql(string sql)
        {
            _migration.Sql.Add(sql);
        }

        public void ElapsedTime(TimeSpan timeSpan)
        {
            _result.ElapsedTime = timeSpan;
        }

        public void Error(string message)
        {
            _result.Errors.Add(message);
        }

        public void Error(Exception exception)
        {
            _result.Errors.Add(exception.ToString());
        }

        public void Complete()
        {
            if (_migration?.Sql != null && _migration.Sql.Any(x => !string.IsNullOrWhiteSpace(x)))
            {
                _result.Migrations.Add(_migration);
            }
        }

        public void Write(string message, bool escaped)
        {
            _migration.Write.Add($"Escaped: {escaped} Message: {message}");
        }
    }
}

```

1.4.0 Migration (ours, not yours)
```
using System.Collections.Generic;

namespace Our.Internal.Models
{
    /// <summary>
    /// A tracking model used to help keep track of what was done
    /// </summary>
    public class Migration
    {
        public Migration()
        {
            Sql = new List<string>();
        }

        /// <summary>
        /// The migration description
        /// </summary>
        public string Description { get; set; }

        /// <summary>
        /// The actual sql commands run by the target migration
        /// </summary>
        public List<string> Sql { get; set; }

        /// <summary>
        /// The type of migration run, such as "up"
        /// </summary>
        public MigrationType MigrationType { get; set; }

        public List<string> Say { get; set; } = new List<string>();

        public List<string> Emphasize { get; set; } = new List<string>();

        public List<string> Write { get; set; } = new List<string>();
    }
}
```

1.4.0 MigrationResult (accumulates migration run results by AppliesTo - multitenancy targets)
```
using System;
using System.Collections.Generic;
using System.Linq;

namespace Our.Internal.Models
{
    /// <summary>
    /// Keep track of which migrations were run for which AppliesTo
    /// </summary>
    public class MigrationResult
    {
        public MigrationResult(AppliesTo appliesTo)
        {
            AppliesTo = appliesTo;
        }

        /// <summary>
        /// Show a summary
        /// </summary>
        public string Summary => !Migrations.Any() || Migrations.All(x => x.Sql.All(string.IsNullOrWhiteSpace))
                    ? "No migrations!"
                    : Errors.Any() ? "There were errors running migrations." : "Success";

        /// <summary>
        /// Track the detail of the migration
        /// </summary>
        public string Detail { get; set; }

        /// <summary>
        /// Track which AppliesTo this was run for
        /// </summary>
        public AppliesTo AppliesTo { get; private set; }

        /// <summary>
        /// Track how long the migration took to run (so we can do analytics)
        /// </summary>
        public TimeSpan ElapsedTime { get; set; }

        /// <summary>
        /// Track which errors we got back. 
        /// </summary>
        public List<string> Errors { get; set; } = new List<string>();

        /// <summary>
        /// The migrations run for this AppliesTo.
        /// </summary>
        public List<Migration> Migrations { get; set; } = new List<Migration>();
    }
}
```

All of which eventually was consumed like so: 
```
<snip>
                    var announcer = new MigrationAnnouncer(response);
                    var runner = new RunnerContext(announcer)
                    {
                        ApplicationContext = string.Empty,
                        Connection = parameter.Tenant.ConnectionString.ConnectionString
                    };
                    var processor = new SqlServerProcessor(
                        parameter.Tenant.ConnectionString,
                        new SqlServer2014Generator() { },
                        announcer,
                        new ProcessorOptions() { PreviewOnly = parameter.WhatIf },
                        new SqlServerDbFactory() { }
                    );

                    var migrationRunner = new FluentMigrator.Runner.MigrationRunner(assembly, runner, processor);

                    try
                    {
                        migrationRunner.MigrateUp();
                    }
<snip>
```

I think that's enough context to show how we were using it in 1.4.0, and I would just like to basically do the same thing in 3.x and later, without relying on deprecated structures, because clearly that's a bad design.

If you say "nah, implement the above and use the `AnnouncerFluentMigratorLogger` then I can do that, it just seems like that's the wrong way to go here.

What did I not remember to provide here to help explain this more clearly at 3am on a Saturday? 😬

**Comment by jzabroski** [comment]
@jcolebrand I read your comments and my heart goes out to you.

@fubar-coder removed the `IAnnouncer` interface because it wasn't clear it provided any benefits over using Microsoft .NET Core LoggerFactory approach.  (That said, it arguably **could** provide some benefits - but the interface just wasn't all that well designed in 1.4.0, so I agree with the decision to remove it.  For an idea of a better Announcer interface, see [Its.Log](https://github.com/jonsequitur/Its.Log) project on GitHub written by a Microsoft employee.  It abstracts away _what to log_ from _how it is logged,_ which is what the announcer interface should have done in the first place.  But the Announcer interface didn't really add much benefit.  As an example, you might want to dial up or down the verbosity on the logging, such that we log rows affected for a delete statement, the names of primary keys, etc.  The current .NET core inspired logging just assumes you use convention-over-configuration for everything and so its fairly spartan in its log messages.)

Anyway, if you're on 1.4, it sounds like your team has not upgraded FluentMigrator in a very long time.  I would probably recommend throwing out whatever you have rather than porting it.  Start from fresh requirements and just write what you need to write.  You can look at FluentMigrator.DotNet.Cli for inspiration for creating an in-process runner.  I have been meaning to complete #1208 but real life gets in the way.  (It would be a really welcome PR to do that.  We can discuss it if you have any doubts on the approach.)

**Comment by fubar-coder** [comment, marked as answer]
First, the `IAnnouncer` is (IIRC) still fully functional even though the output might not be 100% compatible, but there's still a class that isn't deprecated yet and which might be something you can build on: [`FluentMigratorLogger`](https://github.com/fluentmigrator/fluentmigrator/blob/6a56b8e6749ab73910d5132bca9f2b5a398c21c8/src/FluentMigrator.Runner.Core/Logging/FluentMigratorLogger.cs).

So, you can still register your own [`ILoggerProvider`](https://docs.microsoft.com/de-de/dotnet/api/microsoft.extensions.logging.iloggerprovider?view=dotnet-plat-ext-6.0), which returns your custom (derived) implementation of `FluentMigratorLogger`. This class is indeed the foundation of the `IAnnouncer` compatibility :wink: . You can find an [example for the logging provider in the sources](https://github.com/fluentmigrator/fluentmigrator/blob/9c87d698b161b3a1bb596a5624c02891a64459c5/src/FluentMigrator.Runner/Logging/LegacyFluentMigratorLoggerProvider.cs), which in turn uses [`AnnouncerFluentMigratorLogger`](https://github.com/fluentmigrator/fluentmigrator/blob/6a56b8e6749ab73910d5132bca9f2b5a398c21c8/src/FluentMigrator.Runner.Core/Logging/AnnouncerFluentMigratorLogger.cs).

If I were you, I'd just copy the code, remove the `Obsolete` attribute, register the logger provider and go with it.

**Comment by jcolebrand** [comment]
Thanks @fubar-coder that's what I was considering doing, but since it's marked obsolete I didn't want to rely on something that would go away.

I'll have a look at what @jzabroski linked to on 1208 and see if I see something good there too.

Maybe I can figure out a way to non-disclosure acceptable show you guys what I'm doing, see if anything jumps out at you.

Cheers!

**Comment by fubar-coder** [comment]
The `FiuentMigratorLogger` isn't going away (at least I didn't plan to remove it and it's not marked as obsolete), so it's safe to use. BUT even when it may be removed in a future version, you can still copy the logger code into your project. The log ID's are extremely unlikely to change (if ever).

**Comment by jcolebrand** [comment]
Honestly it's just the IAnnouncer that we wanted to hook. I gotta figure out how to non-disclosure acceptable way to share what I'm doing. You would absolutely hate it I'm sure :D :D

**Comment by jzabroski** [comment]
#1208 is really about making it easier to extend FluentMigrator in a forward-looking way.  A lot of people basically create their own runner logic, and it would make more sense if the main bits were just there and easy to extend but with custom DI.  It's not about Announcers, per say.  It's really about, OK, I want to customize FluentMigrator with my own conventions, my own logging, etc. What's the best way to do that so that everyone has a clear way to do that?  I think OOP inheritance is probably the cleanest way to do that in C#, especially for "application frameworks" like FluentMigrator, as it allows to customize the run-time "object graph" at arbitrary points in the object graph via constructor injection.

**Comment by jzabroski** [comment]
@jcolebrand I saw your reply in Gitter.  It's better to have a discussion with GitHub Discussions.  We try to discourage Gitter moving forward as it duplicates the primary purpose of GitHub Discussions, and with GitHub discussions, we can cross-link those chats to issues all in one platform.

**Comment by jzabroski** [comment]
@jcolebrand said on Gitter:

> @jzabroski is this a good place to chat about fluentmigrator/fluentmigrator#1556 cos I have a specific question about "have you guys already discussed X elsewhere that I can see history"? Specifically, I'm not sure where to look for "have you guys ever returned an IList<AnnouncerResult> (for lack of a better model name) on the MigrateUp/MigrateDown to mirror what is done on the log extensions utility but as a return object. If you never consume it, then it just gets thrown away, and the runtime handles that, but in our case that solves the need. Honestly I'm curious where any discussions would have occurred on that and where to look for those, because if it hasn't been a case considered before, I'm happy to open a discussion/feature request (and I can help put in a PR to plumb it in if that's desired once it's deemed a worthwhile thing to add) - Hence wanting to figure out if it's been discussed before so I can know the history.
> I'm not sure who else I might tag here besides John, I just know he seems to be a key person to start with from responses I've seen on issues

**Comment by jcolebrand** [comment]
@jzabroski Oh, whew, I thought I would be cluttering up Github Issues if I did that, and wasn't sure which way to go. Much happier to have that conversation here.

Do you have any thoughts on the above? Happy to read over any previous discussions, and/or submit a PR that sort of shows what I mean on the overall.

Cheers!

**Comment by jzabroski** [comment]
@jcolebrand Specifically, you are asking about this, I think?

> have you guys ever returned an IList (for lack of a better model name) on the MigrateUp/MigrateDown to mirror what is done on the log extensions utility but as a return object.

If so, I think what you are actually asking is, "Is there any way I can get a list of migration expressions run in each MigrateUp/MigrateDown call?"  The answer is, Yes, it's possible. But also: it's not the nicest API, IMHO.  @fubar-coder actually added a unit test to help me with this awhile ago. https://github.com/fluentmigrator/fluentmigrator/commit/20e806f4171e0aa69bed92f43d898bea80f40c36

The downside is that, since expressions are collected in the context (because the context is what ultimately is the "execution context" for a set of migrations), you can't just intercept GetUpExpressions().  I am guessing the reason you would want to intercept them is so that you can log them.  However, that could lead to some funky-looking log messages, because you would be logging _the promise to execute them_ rather than the _execution itself_.

**Comment by jcolebrand** [comment]
Yeah, that's basically it. What I've done instead is also clunky, I'll share code here so you can see. I pulled the repo and think I've got most of what you guys are doing, which is why I think it could be managed to return something like what I'm looking for, but I get most consumers probably don't care either. Given how we utilize the framework, we run our migrations in parallel across all our tenants even if they've already had them in the past, due to tenants being moved around in our multitenant environment. And we don't do the migrations on startup of app, we run them in an out-of-band (separate app and database run permissions, yeah?)

Part of my solution involves loading the FluentMigrator DLLs on demand because my app can load either 1.4, 2.x, 3.x (or if we load from another migrator, which we don't do yet, but we could) all via one console driver, according to what was compiled against (I do a lot of other work, including downloading stuff, validating internal manifests, some other stuff)

So here's my 3.3.1 logger implementation that I'm using now: 
```
namespace SRE.MigrationUtility.Core.FluentMigrator.MSSQL331
{
    public class FluentMigratorNewLogger : ILogger
    {
        public FluentMigratorNewLogger(
            IAppliesToResolver appliesTo
        )
        {
            _migrationResult = new MigrationResult(appliesTo.AppliesTo);
            _migration = new Migration("Preliminary logged details");
        }

        private MigrationResult _migrationResult { get; }
        private Migration _migration;

        public IDisposable BeginScope<TState>(TState state) { throw new NotImplementedException(); }
        public bool IsEnabled(LogLevel logLevel) => true;

        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
        {
            if (null == _migrationResult) { throw new Exception("you did not initialize the migration result"); }

            var message = null != formatter ? formatter(state, exception) : state.ToString();

            if (new[] { "COMMIT TRANSACTION", "ROLLBACK TRANSACTION" }.Contains(state as string))
            {
                Console.WriteLine(message);
            }

            if (exception != null)
            {
                _migrationResult.Error(exception.Message);
            }

            if (eventId.Id == RunnerEventIds.Sql.Id)
            {
                var sql = message;
                if (string.IsNullOrEmpty(sql)) { sql = "No SQL statement executed."; }
                _migration.Sql(sql);
            }
            else if (eventId.Id == RunnerEventIds.Emphasize.Id)
            {
                _migration.Emphasize(message);
            }
            else if (eventId.Id == RunnerEventIds.Heading.Id)
            {
                if (_migration != null)
                {
                    _migrationResult.Migration(_migration);
                }
                _migration = new Migration(message);
            }
            else if (eventId.Id == RunnerEventIds.ElapsedTime.Id)
            {
                if (typeof(TState) == typeof(TimeSpan))
                {
                    _migrationResult.ElapsedTime = (TimeSpan)(object)state;
                }
            }
            else
            {
                _migration.Say(message);
            }
        }

        public MigrationResult Complete()
        {
            if (null != _migration)
            {
                _migrationResult.Migration(_migration);
            }

            return _migrationResult;
        }
    }

    public class FluentMigratorNewLogger<T> : FluentMigratorNewLogger, ILogger<T>
    {
        public FluentMigratorNewLogger(IAppliesToResolver appliesTo) : base(appliesTo) { }
    }
}
```

And it's being used like this (edited for brevity):

```
                serviceCollection
                    .Remove(typeof(ILogger))
                    .Remove(typeof(ILogger<>))
                    .AddScoped(typeof(IAppliesToResolver), sp => new AppliesToResolver(parameter.Tenant))
                    .AddScoped<ILogger, FluentMigratorNewLogger>()
                    .AddScoped(typeof(ILogger<>), typeof(FluentMigratorNewLogger<>));

                using (var scope = provider.CreateScope()) {
                        var runner = scope.ServiceProvider.GetService<IMigrationRunner>();
                        runner.MigrateUp();
                        migrationResult = ((FluentMigratorNewLogger)runner.GetType().GetField("_logger", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(runner)).Complete();
                }
```

I think you can get the gist of what I'm doing with these snippets here. It's not _great_ but it gets the job mostly done for what I'm doing. Basically, anytime you call the LoggerExtensions in https://github.com/fluentmigrator/fluentmigrator/blob/master/src/FluentMigrator.Runner/MigrationRunner.cs or the related VersionInfoMetadataTable creation is what I want to capture, as our processes don't typically ILogger in them (if they did, they would just get captured to my "standard" custom logger elsewhere, so I would still capture the output. (I log to either raw console output or to TeamCity appropriately wrapped output via my own LoggingExtension classes, as you do)

I'm actually really curious how many people have written something like this over time, we can't be that unique. There may be a better way to return (see the use of my Complete() method, and the injected tenant record for the multienancy header information)

I can show the classes for the other two objects if you want, but they are just collections of lists and some status message captures, nothing fancy or special. You can basically reconstruct them from what you see here and we would have a close parity I'm sure.

As I read through the source code tracking the IAnnouncer and ILogger deltas, it definitely looks like the IAnnouncer isn't gonna do what I want anymore. Thoughts @fubar-coder ?

**Comment by jcolebrand** [comment]
And I should also like to repeat @jzabroski that I get that adding code to support my specific use-case would be silly, so that's why I was curious if there had been a similar previous discussion so I could read up on it and understand prior development efforts. I could see it as useful even if there were unlikely to be many consumers in the future. Since C# just throws away uncaptured return objects, and since mine is relatively lightweight (model of List<T>s) it's not terribly high overhead, but if I'm the lone asker, then I should probably just work on a better version of my implementation for me.

**Comment by fubar-coder** [comment]
@jcolebrand We don't have a monitoring or statistics collection infrastructure, and the IAnnouncer seemed just for writing logs. We (or @jzabroski specifically) might consider adding this functionality, as it seems useful, but as I wrote: Regardless of IAnnouncer being obsolete, you can and will in the forseeable future be able to use FluentMigratorLogger and the logger IDs.

**Comment by jcolebrand** [comment]
Yeah, I'm definitely relying on the logger ids to capture what I want. Just can't let it go straight to console or whatnot, and maybe I missed how I should be implementing or utilizing FluentMigratorLogger on rereading this. Is it supposed to be inherited and captured from @fubar-coder ?

**Comment by fubar-coder** [comment]
@jcolebrand That is correct. It's used to be inherited. There are functions/methods that you can override to capture the given messages, so that you don't have to rely on the log IDs, but **maybe** there is a way to implement filters and maybe there's a way to filter/collect the log messages? That might be an even better approach.

**Comment by jzabroski** [comment]
@jcolebrand I think what you are describing is logging "activities" rather than "events".  Time interval statistics are a property of an activity.  An event is a point in time.  That's why I referenced Its.Log as an example of a better way to do announcements.  Sometimes, announcements are about activities rather than events.

The way Its.Log (and even Serilog, honestly) work is you forward log data, unformatted, to a formatting engine, that formats it and logs it.  The idea is to log objects, not strings.  The downside to such approaches, of course, is if consumers virtualized their ToString implementation and call a lock inside of it, you could get a deadlock.  But I have always felt the overall approach was smooth.  I just wished there was an analyzer that warned consumers about logging stuff in unsafe ways.

**Comment by jcolebrand** [comment]
Then I may have dismissed the Its.Log too soon for not understanding the full implication of what it can offer, I will go back and investigate it further @jzabroski 

We are running these in parallel using TPL and PLINQ per target environment, so they are needed in aggregate per environment, so we can do reporting per as needed.

O(1) validate inputs -> O(1) do some pre-work re: library locations, engine -> O(1) other pre-work -> O(n) parallel invoke FM -> O(1) reporting on FM activities in a consolidated fashion

If I were to write-to-disk even with Serilog, I would still have issues of having to reimport and reclassify those log events into the activities. The approach I have now correlates the activities with the events, right? Or am I misunderstanding you still?

Let me do some reading on Its.Log to see if it effectively helps to encapsulate what I'm trying to accomplish.

**Comment by jzabroski** [comment]
> If I were to write-to-disk even with Serilog, I would still have issues of having to reimport and reclassify those log events into the activities. The approach I have now correlates the activities with the events, right? Or am I misunderstanding you still?

I think I understand what you are getting at.

I had some machinations of a powerful way to mock and test migrations, whereby you could seed a MigrationContext with a database state that is basically an ASN.1 dictionary of object paths.  For example, you can think of the output of `SELECT * FROM INFORMATION_SCHEMA.COLUMNS` ANSI SQL compatibility view as a series of ASN.1 paths to an object, where the object is a column.  So, if you then wrote `Delete.Column("x").FromTable("foo").InSchema("bar");` we could index into that dictionary to see whether the metadata exists or not.  If it exists, then its a trivial precondition check to say the delete column operation should succeed.

I had thought about such stuff, but didnt have the chance to implement it.  The main difference between what you are saying and I am saying, is it would be beneficial if there was a standard metadata format for common activities FluentMigrator performs, and I can use that standard metadata to write validation rules.  (Maybe you are saying this, but you didnt give an exact example like my delete statement).  The upshot of all this is the connectionless runner could be run against a ASN.1 snapshot of a production database.  That would be insanely useful for environments where DBAs insist on a RunAllMigrations.sql file be generated, because we could guarantee (some) correctness properties of the generated file.

Probably easier to explain further on a Zoom call if you're interested in doing all this.

Edit: Also, ASN.1 is just one possible way to think about storing object paths.  An IDL (Interface Definition Language) type data structure would also work well.  For example, you might want to change a `text` data type to `nvarchar(max)`, so you need a way to store not just the column name, but the data type and it's nullability as well.

**Comment by jcolebrand** [comment]
Oh, yeah, no, I wouldn't be doing something like _that_ for logging checking. I see where you're going with that now re: activities, which isn't quite what I'm looking for. I just want to know basically when migrations finish and against what metadata did they finish. Those activity graphs on the ASN.1 would be intense, and more than I suspect most of what would be useful to me for my goals.

It's more about, like, uhhh ... consider this migration class (where ASN.1 wouldn't even come into play as no actions have occurred on the target database, but a record marker would certainly show up in the metadata table to prevent it being rerun, which is the principle use-case of FM for us -- most of our runs are )

```
using System.Threading;
using FluentMigrator;

namespace Framework.Multiple.Migration1
{
    [Migration(1, "First")]
    public class Class1 : Migration
    {
        public override void Down()
        {
            Thread.Sleep(RandomTimeSpan.Get());
        }

        public override void Up()
        {
            Thread.Sleep(RandomTimeSpan.Get());
        }
    }
}
```

And here's the implementation for that static class

```
using System;

namespace Framework.Multiple.Migration1
{
    public static class RandomTimeSpan
    {
        public static Random Random = new Random();

        public static TimeSpan Get()
        {
            return new TimeSpan(0, 0, Random.Next(1,10));
        }
    }
}
```

I want to, in parallel, run that across 10 target databases, and collect the information about which of the 10 databases had the migration applied _at the runtime_ (I could, in fact, interrogate the databases for the applied date, but I already would have gotten that from the existing run, and I would lose any ancillary data that may have been generated, such as runtime per migration applied, connection string of the target environment, etc). Did it get applied to all 10 databases? To only one of them? The default console logger gives me almost none of the metadata, and definitely doesn't tell me the target that I was writing to/migrating.

Now, this is a highly contrived example, so of course it's not really useful for a _lot_ of information. But let's say that I had 10 real migrations to each apply to 10 target databases, but of those, 3 had already received all the migrations in the set, 2 had only received 7 of the previous migrations, 3 had only received 5 of the previous migrations, and 2 had received none of the migrations.

Again, we are doing this in parallel as a metarunner but in a single controlling process (hence the O(1) parts of the above call chain). So we would take as long as the 10 migrations to run, but 3 of the databases would effectively end up with "List<migrationresult>.Count = 0" as it were, and the others at 3 applied and 5 applied, etc. 

If I only wrote those to console, or to Serilog -> file, then I don't have a way to batch them by target database, only by runtime. So I could reprocess those logs somehow, and provided the record told me the metadata (which connection string got the migration?) I could reconstruct my results graph.

But by using the approach above, I'm keeping all of that in memory, so I can just work with an in-memory list of results.

A zoom call would probably go a long way to you understanding in depth more of what I'm trying to accomplish. I would love to just throw it all up in a github repo somewhere but I know the company would not appreciate that. The easiest gmail to reach me at is the same as this handle, but it also gets a lot of other regular mail, so I miss unexpected emails. (in this case it would be expected tho, yeah? but future readers may not realize that if they tried to reach out to me)

Our codebase has thousands of migration classes across hundreds of projects, and when I say I'm applying them in parallel across a number of target databases, I do mean in some cases 250+ databases in a single push, with as many as 30-40 active migrations per target invocation, with hundreds of dlls being migrated from in the course of a deploy window on the worst of times. Being able to aggregate the data in-situ is ideal for me, much more so than trying to hope I can parse console.writeline results in realtime back into events in the call stream. And clearly doing (250 * 30) in serial takes much longer than (250 * 30 / parallelization) just so I can get the results from log per run in a clean rebuildable fashion.

Again, ASN.1 wouldn't apply, insofar as I can tell from your example above, as this is what we generally end up with on our platform:

```
using CompanyName.Migrations;
using FluentMigrator;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CompanyName.SomeProjectName.Migrations.Migrations.DEV_121345
{
    [ServiceMigration("DEV-121345", 2021, Month.July, 26)]
    public class MigrateSomeClassDescriptiveName : Migration
    {
        public override void Down()
        {
            Execute.EmbeddedScript("Dev-121345.down.sql"); // Needs to be embedded in the dll
        }

        public override void Up()
        {
            Execute.EmbeddedScript("Dev-121345.up.sql"); // Needs to be embedded in the dll
        }
    }
}
```

So in this case I would be getting the descriptive information from the metadata (they like to have their own overriding attributes that implement over your attribute class to make it easier for the teams to parse what becomes a YYYYMMDD style database identifier on the base migration attribute, with description, etc) and I would pick up whatever else gets logged today (cos I did originally share the IAnnouncer codepath from 1.4.0 that we use in production way up above) but clearly you would have no way to infer the actual operations on the database to generate any of the ASN.1 type data that you referenced.

If you want to ping me on email and we can set something up, I'm happy to do so. I know we all have busy schedules, but I'm pretty flexible on when I can chat. We can definitely nail something down, I'm sure.

And like I say, if you guys are open to seeing something like what I'm proposing I am more than happy to do the work to setup the PR, just not sure _how_ useful it would be for anyone who isn't me. I'm sure we aren't the only people doing something this strange, but are there 10 of us? 100? 🤷 

Thanks again fellas!
